Update Kafka connector to 2.3.1

dcfocus opened the pull request:
This PR improves Kafka connector implementation and performance.

1. Replaced SimpleConsumer with KafkaConsumer for Kafka connector.
2. Push collecting offsets for each partition from coordinator to workers to resolve the bottleneck.

dcfocus commented:
@rongrong @zhenxiao 
Refactored previous PR, this one is the main one with upgrading Kafka client. Other features will come later, including nested ROW type support and zookeeper discovery, etc.

beinan commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaThreadIdentifier.java:
```diff
@@ -0,0 +1,62 @@
+{
+    public final String threadId;
+    public final HostAddress hostAddress;
+    public final String partitionId;
```
The partitionId is always a number, isn't it? why using a string type here?

dcfocus replied:
Good point. 

Also partitionId info was used as debug info in log.debug(). I don't think it's required here. Removed it for simplification.

beinan commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConsumerManager.java:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    public KafkaConsumer getConsumer(KafkaThreadIdentifier consumerId)
```
We are not really using this method, looks like we create a new consumer every time in each thread.  If so, I think we don't need the Loading Cache.

dcfocus replied:
Good point. Removed the unused cache.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaThreadIdentifier.java:
```diff
@@ -0,0 +1,62 @@
+            return false;
+        }
+        final KafkaThreadIdentifier other = (KafkaThreadIdentifier) obj;
+        return Objects.equals(this.threadId, other.threadId);
```
KafkaThreadIdentifier class might be redundant. Removed.

dcfocus replied:
KafkaThreadIdentifier might not be necessary. Removed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -68,7 +69,7 @@
 
     KafkaRecordSet(KafkaSplit split,
-            KafkaSimpleConsumerManager consumerManager,
+                   KafkaConsumerManager consumerManager,
```
incorrect format

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
         ImmutableList.Builder<Type> typeBuilder = ImmutableList.builder();
 
-        for (DecoderColumnHandle handle : columnHandles) {
+        for (DecoderColumnHandle handle : this.columnHandles) {
```
Presto generally does not use this to reference class variables, just use columnHandles

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
 
         this.columnTypes = typeBuilder.build();
+        KafkaConsumer consumer = consumerManager.createConsumer(Thread.currentThread().getId(), split.getLeader());
+        TopicPartition tp = new TopicPartition(split.getTopicName(), split.getPartitionId());
```
s/tp/topicPartition/g

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+        TopicPartition tp = new TopicPartition(split.getTopicName(), split.getPartitionId());
+        consumer.assign(ImmutableList.of(tp));
+        setOffsetRange(consumer, split);
+        consumer.close();
```
calling close in constructor does not seem good
how about make KafkaRecordSet closable?

dcfocus replied:
Completed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+        consumer.close();
+    }
+
+    private static void setOffsetRange(KafkaConsumer<Long, String> consumer, KafkaSplit split)
```
s/setOffsetRange/setOffset/g

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+
+        long beginningOffset =
+                (split.getStartTs() == 0) ?
+                        consumer.beginningOffsets(ImmutableList.of(topicPartition)).values().iterator().next() :
```
could we use KafkaConsumer#position() to get the offset?

dcfocus replied:
position() returns next record that will be fetched for a TopicPartition. Here we want first offset available on the partition. position() doesn't work in this scenario.

dcfocus replied:
Here I want to do is, if no time range is specified by user, then we should fetch the entire topicPartition.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+        consumer.assign(ImmutableList.of(topicPartition));
+
+        long beginningOffset =
+                (split.getStartTs() == 0) ?
```
formatting: move (split.getStartTs() == 0) ? to the previous line

dcfocus replied:
fixed

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+        split.setEnd(endOffset);
+    }
+
+    private static long findOffsetsByTimestamp(KafkaConsumer<Long, String> consumer, TopicPartition tp, long timestamp)
```
please not use variable abbreviation:
s/tp/topicPartition/g
s/tpOffsets/topicPartitionOffsets/g

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+            if (tpOffsets == null || tpOffsets.values().size() == 0) {
+                return 0;
+            }
+            OffsetAndTimestamp offsetAndTimestamp = new ArrayList<>(tpOffsets.values()).get(0);
```
better not create new ArrayList. Could we use Collections directly?

dcfocus replied:
Good point. Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+                return 0;
+            }
+            OffsetAndTimestamp offsetAndTimestamp = new ArrayList<>(tpOffsets.values()).get(0);
+            return (offsetAndTimestamp == null) ? 0 : offsetAndTimestamp.offset();
```
is it possible offsetAndTimestamp is null? Or let's make it Optional

dcfocus replied:
I check the code, look like it could be `null`. I will use Optional
`      offsetsByTimes.put(entry.getKey(), (Object)null);`

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+            OffsetAndTimestamp offsetAndTimestamp = new ArrayList<>(tpOffsets.values()).get(0);
+            return (offsetAndTimestamp == null) ? 0 : offsetAndTimestamp.offset();
+        }
+        catch (Exception e) {
```
Exception is too wide, is it TimeoutException, AuthenticationException, or UnsupportedVersionException? Let's catch them explicitly.

dcfocus replied:
Looks like `IllegalArgumentException` is the only thrown Exception type from try block. Will explicitly catch it.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -84,11 +85,54 @@
+        }
+        catch (Exception e) {
+            log.error(e, String.format("Failed to find offset by timestamp: %d for partition %d", timestamp, tp.partition()));
+            throw new RuntimeException(e);
```
We should throw IOException, or create a Kafka connector exception

dcfocus replied:
Good point! Added a `KAFKA_CONSUMER_ERROR` exception.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -305,30 +353,28 @@ private void checkFieldType(int field, Class<?> expected)
         public void close()
         {
+            log.debug("Close consumer if exist: %d, partition %d", Thread.currentThread().getId(), split.getPartitionId());
+            if (this.consumer != null) {
```
not use this to reference class members

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -305,30 +353,28 @@ private void checkFieldType(int field, Class<?> expected)
         @Override
         public void close()
         {
+            log.debug("Close consumer if exist: %d, partition %d", Thread.currentThread().getId(), split.getPartitionId());
```
this log might not be needed

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -305,30 +353,28 @@ private void checkFieldType(int field, Class<?> expected)
                     }
 
-                    messageAndOffsetIterator = fetchResponse.messageSet(split.getTopicName(), split.getPartitionId()).iterator();
+                    TopicPartition tp = new TopicPartition(split.getTopicName(), split.getPartitionId());
```
variable abbreviation, s/tp/topicPartition/g

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -305,30 +353,28 @@ private void checkFieldType(int field, Class<?> expected)
 
-                    messageAndOffsetIterator = fetchResponse.messageSet(split.getTopicName(), split.getPartitionId()).iterator();
+                    TopicPartition tp = new TopicPartition(split.getTopicName(), split.getPartitionId());
+                    this.consumer.assign(ImmutableList.of(tp));
```
remove this for class member reference

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -305,30 +353,28 @@ private void checkFieldType(int field, Class<?> expected)
+                    TopicPartition tp = new TopicPartition(split.getTopicName(), split.getPartitionId());
+                    this.consumer.assign(ImmutableList.of(tp));
+                    this.consumer.seek(tp, cursorOffset);
+                    ConsumerRecords<ByteBuffer, ByteBuffer> records = this.consumer.poll(3000);
```
make 3000 a constant
is 3s too long for a poll?

dcfocus replied:
Fixed. Reduced it to 1s, which should be good enough from the current performance data.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -305,30 +353,28 @@ private void checkFieldType(int field, Class<?> expected)
+                    this.consumer.assign(ImmutableList.of(tp));
+                    this.consumer.seek(tp, cursorOffset);
+                    ConsumerRecords<ByteBuffer, ByteBuffer> records = this.consumer.poll(3000);
+                    this.consumer.commitSync();
```
why need commitSync?

dcfocus replied:
Good point. It's not required. I copied over from a consumer example.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplit.java:
```diff
@@ -55,8 +57,8 @@ public KafkaSplit(
             @JsonProperty("partitionId") int partitionId,
-            @JsonProperty("start") long start,
-            @JsonProperty("end") long end,
+            @JsonProperty("startTs") long startTs,
```
s/startTs/startTimestamp/g
s/endTs/endTimestamp/g

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplit.java:
```diff
@@ -42,8 +42,10 @@
     private final int partitionId;
-    private final long start;
-    private final long end;
+    private long start;
```
move non-final variables below final variables

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplit.java:
```diff
@@ -160,8 +186,8 @@ public String toString()
                 .add("keyDataSchemaContents", keyDataSchemaContents)
                 .add("messageDataSchemaContents", messageDataSchemaContents)
                 .add("partitionId", partitionId)
-                .add("start", start)
```
we still need start end for toString, right?

dcfocus replied:
Both start and end are initialized with default value of `0` when splits are built. 
They are set with right values after consumer talking to partitions on Presto workers through API getOffsetsByTimeStamp(). 

To my understanding, toString is used for serialization and Deserialization when splits are passed among workers. Since setting the two variables happens locally on a Presto worker. We don't need them in toString. 

But I think you are right to keep them for clarity and simplicity. Added them back.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java:
```diff
@@ -93,44 +87,38 @@ public ConnectorSplitSource getSplits(
         KafkaTableHandle kafkaTableHandle = convertLayout(layout).getTable();
         try {
-            SimpleConsumer simpleConsumer = consumerManager.getConsumer(selectRandom(nodes));
+            HostAddress node = this.servers.selectRandomServer();
```
remove this

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java:
```diff
@@ -93,44 +87,38 @@ public ConnectorSplitSource getSplits(
-            TopicMetadataRequest topicMetadataRequest = new TopicMetadataRequest(ImmutableList.of(kafkaTableHandle.getTopicName()));
-            TopicMetadataResponse topicMetadataResponse = simpleConsumer.send(topicMetadataRequest);
+            KafkaConsumer<String, String> consumer = consumerManager.createConsumer(Thread.currentThread().getId(), node);
+            List<PartitionInfo> parts = consumer.partitionsFor(topic);
```
s/parts/partitions/g

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java:
```diff
@@ -93,44 +87,38 @@ public ConnectorSplitSource getSplits(
+                long startTs = layoutHandle.getOffsetStartTs();
+                long endTs = layoutHandle.getOffsetEndTs();
+
+                KafkaSplit split = new KafkaSplit(
```
could you please elaborate more on why pushdown offsets computation for split?
could we compute offsets here for all splits, if the computation overhead is the same? As Coordinator decouples split generation with split scan

dcfocus replied:
Before pushdown, offset computation is done split by split in serial in a big for loop, by a single thread I believe. 
If we delay offset computation to after splits are distributed to workers, this computation can be done in parallel by multiple threads on multiple workers. 
Basically we gain performance benefit by parallelizing the work.

zhenxiao replied:
how expensive is offset computation? Do you have benchmarks shows how much performance gain we get by pushdown offset computation?
The logic is a more complicated after pushdown. we could do the pushdown if get big performance gain.

dcfocus replied:
Here are some numbers. For  a topic with 1000 partitions, it takes average of 40120 millisecond to build all 1000 splits before the pushdown. It's done on a single machine (the coordinator).
After pushdown, this can be finished for 1 millisecond.
Above number are average from 10 runs. 

The gain is about a fixed 41 seconds. Whether it's worth it really depends on the use case, and cluster settings. For example, if we are interested in 1-minute data from Kafka, then we can reduce 1 minute execution time to couple seconds. But if we are interested in 1-hour data, then it's saving 40 seconds from several minutes. 

I will leave this decision to you.

zhenxiao replied:
thank you, Da. I did some research, seems most Kafka topics has < 100 partitions, Linkedin has average ~70 partitions per topic: https://conferences.oreilly.com/strata/strata-ca-2018/public/schedule/detail/63921
Am inclined to the following:
at a first step, have an implementation without pushdown, as most users will not have hundreds of topic partitions, and the performance penalty is not big. 
We could file a following improvement PR, to discuss the details of pushdown implementation. (the current pushdown logic is kind of complicated)
what do you think?

dcfocus replied:
Sound good. I will update the PR to remove it.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaTableLayoutHandle.java:
```diff
@@ -36,9 +48,24 @@ public KafkaTableHandle getTable()
-        return table.toString();
+        return toStringHelper(this)
+                .add("table", table.toString())
+                .add("offset_start_ts", offsetStartTs)
```
s/offset_start_ts/startOffsetTimestamp/g
generally, we do not put underscore in toString

dcfocus replied:
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConnectorConfig.java:
```diff
@@ -69,6 +62,16 @@
+    /**
+     * Maximum number of bytes from one partition per poll()
+     */
+    private int maxPartitionFetchBytes = 1048576;
```
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConsumerManager.java:
```diff
@@ -0,0 +1,68 @@
+    {
+        requireNonNull(kafkaConnectorConfig, "kafkaConfig is null");
+        this.maxPartitionFetchBytes = kafkaConnectorConfig.getMaxPartitionFetchBytes();
+
```
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConsumerManager.java:
```diff
@@ -0,0 +1,68 @@
+    {
+        final Properties props = new Properties();
+        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+                hostAddress.toString());
```
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaTableLayoutHandle.java:
```diff
@@ -17,17 +17,29 @@
 {
     private final KafkaTableHandle table;
 
+    // Start timestamp used to seek to the start offset
```
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaTableLayoutHandle.java:
```diff
@@ -17,17 +17,29 @@
+    private final long offsetStartTs;
+
+    // End timestamp used to seek to the end offset
+    private final long offsetEndTs;
```
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaMetadata.java:
```diff
@@ -204,7 +207,27 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     {
         KafkaTableHandle handle = convertTableHandle(table);
-        ConnectorTableLayout layout = new ConnectorTableLayout(new KafkaTableLayoutHandle(handle));
+        Long startTs = null;
```
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaMetadata.java:
```diff
@@ -204,7 +207,27 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
         KafkaTableHandle handle = convertTableHandle(table);
-        ConnectorTableLayout layout = new ConnectorTableLayout(new KafkaTableLayoutHandle(handle));
+        Long startTs = null;
+        Long endTs = null;
```
Fixed.

dcfocus commented:
Hi @zhenxiao Thanks for all the comments. Very helpful!!
I fixed most of the comments and left explanation for the rest. 
Can you take a look once you get time?
(Current build failed due to time exception of 1 hour 20 min. Will pass if re-run. But seems Travis doesn't allow re-run. )

dcfocus commented on pom.xml:
```diff
@@ -1048,6 +1048,30 @@
+            <dependency>
+                <groupId>org.apache.kafka</groupId>
+                <artifactId>kafka-clients</artifactId>
+                <version>1.1.1</version>
```
Was to keep consistent with `kafka_2.12`

cricket007 replied:
I'm not sure there's a correlation between the Scala version and Kafka version.

If Presto needs Scala 2.12, that's fine, but I mean, why not target Kafka 2.3.0, the latest version, for example?

dcfocus replied:
Thanks Jordan for explanation. Yes, you are right. There shouldn't be correlation. I will update to latest compatible version.

zhenxiao replied:
same question, shall we use kafka 2.3.1, the latest version?

dcfocus replied:
Right, but I keep running into this issue after I tried with 2.0.0+
`Caused by: java.lang.ClassNotFoundException: org.apache.kafka.common.record.RecordFormat`

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConsumerManager.java:
```diff
@@ -0,0 +1,64 @@
+        this.maxPollRecords = kafkaConnectorConfig.getMaxPollRecords();
+    }
+
+    KafkaConsumer<String, String> createConsumer(Long threadId, HostAddress hostAddress)
```
Good catch. Thanks for pointing out.

dcfocus commented:
> This PR improves Kafka connector implementation and performance.
> 
> 1. Replaced SimpleConsumer with KafkaConsumer for Kafka connector.
> 2. Push collecting offsets for each partition from coordinator to workers to resolve the bottleneck.

Removed change #2 after discussion with @zhenxiao . For most companies, kafka topic is not that big, we may not have significant performance gain from pushdown of getOffsetByTimeStamp.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java:
```diff
@@ -93,44 +87,40 @@ public ConnectorSplitSource getSplits(
 
-            TopicMetadataRequest topicMetadataRequest = new TopicMetadataRequest(ImmutableList.of(kafkaTableHandle.getTopicName()));
-            TopicMetadataResponse topicMetadataResponse = simpleConsumer.send(topicMetadataRequest);
+            KafkaConsumer<String, String> consumer = consumerManager.createConsumer(Thread.currentThread().getId(), node);
```
Fixed.

dcfocus commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -109,13 +146,17 @@ public RecordCursor cursor()
+        private Iterator<ConsumerRecord<ByteBuffer, ByteBuffer>> messageAndOffsetIterator;
         private final AtomicBoolean reported = new AtomicBoolean();
-
+        private KafkaConsumer consumer;
```
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConsumerManager.java:
```diff
@@ -0,0 +1,65 @@
+
+    KafkaConsumer<ByteBuffer, ByteBuffer> createConsumer(Long threadId, HostAddress hostAddress)
+    {
+        final Properties props = new Properties();
```
s/props/properties/g

dcfocus replied:
Fixed.

zhenxiao replied:
s/props/properties/g ?

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConsumerManager.java:
```diff
@@ -0,0 +1,65 @@
+    KafkaConsumer<ByteBuffer, ByteBuffer> createConsumer(Long threadId, HostAddress hostAddress)
+    {
+        final Properties props = new Properties();
+        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, hostAddress.toString());
```
static import these ConsumerConfig constants

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConsumerManager.java:
```diff
@@ -0,0 +1,65 @@
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
+        props.put(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG, "org.apache.kafka.clients.consumer.RoundRobinAssignor");
+
+        Thread.currentThread().setContextClassLoader(null);
```
why set classloader to null?

dcfocus replied:
Fixed. Was there to get ride of some exception during experiments.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -305,30 +308,26 @@ private void checkFieldType(int field, Class<?> expected)
+                    long threadId = Thread.currentThread().getId();
+
+                    if (consumer == null) {
+                        consumer = consumerManager.createConsumer(threadId, split.getLeader());
```
use Thread.currentThread().getName(), instead of creating threadId

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java:
```diff
@@ -55,11 +54,11 @@
 
-    private static final int KAFKA_READ_BUFFER_SIZE = 100_000;
     private static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
+    private static final int POLL_TIMEOUT = 1_000;
```
I still feel 1s is too long, how about 500ms?

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplit.java:
```diff
@@ -42,9 +42,9 @@
     private final Optional<String> keyDataSchemaContents;
     private final Optional<String> messageDataSchemaContents;
     private final int partitionId;
-    private final long start;
```
why change start from final to non-final? We should keep start and end untouched, is it?

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplit.java:
```diff
@@ -83,12 +83,22 @@ public long getStart()
         return start;
     }
 
+    public void setStart(long start)
```
do we need to set start and end?

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java:
```diff
@@ -93,44 +92,35 @@ public ConnectorSplitSource getSplits(
+            HostAddress node = servers.selectRandomServer();
+            String topic = kafkaTableHandle.getTopicName();
 
+            KafkaConsumer<ByteBuffer, ByteBuffer> consumer = consumerManager.createConsumer(Thread.currentThread().getId(), node);
```
use thread getName(), instead of getId()

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java:
```diff
@@ -143,6 +133,41 @@ public ConnectorSplitSource getSplits(
         }
     }
 
+    private static long[] getOffsets(KafkaConsumer<ByteBuffer, ByteBuffer> consumer, PartitionInfo partition, long startTimestamp, long endTimestamp)
```
would it better return a pair of long? instead of having a long array?
Or, we could inline this functions

dcfocus replied:
Good point. Inlined, fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java:
```diff
@@ -93,44 +92,35 @@ public ConnectorSplitSource getSplits(
                 }
+
+                HostAddress partitionLeader = HostAddress.fromParts(leader.host(), leader.port());
+                KafkaTableLayoutHandle layoutHandle = (KafkaTableLayoutHandle) layout;
```
move this line above the for loop

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaStaticServerset.java:
```diff
@@ -0,0 +1,51 @@
+import static java.util.Objects.requireNonNull;
+
+public class KafkaStaticServerset
+        implements KafkaCluster
```
do we need KafkaCluster interface? It only has one implementation
we might just need one class with selectRandomServer()

dcfocus replied:
Fixed.

zhenxiao commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplit.java:
```diff
@@ -42,9 +42,9 @@
     private final Optional<String> keyDataSchemaContents;
     private final Optional<String> messageDataSchemaContents;
     private final int partitionId;
+    private final HostAddress leader;
```
no need to move leader here, is it?

dcfocus replied:
Fixed.

zhenxiao commented:
Hi @cricket007 upon tests passed, am mostly good with this PR
could you please do one more pass before I merge?

cricket007 commented on pom.xml:
```diff
@@ -1038,7 +1038,7 @@
                 <groupId>org.apache.kafka</groupId>
                 <artifactId>kafka_2.12</artifactId>
-                <version>1.1.1</version>
+                <version>2.3.1</version>
```
Could we extract this version as a property?

dcfocus replied:
Fixed.

cricket007 commented on presto-kafka/src/test/java/com/facebook/presto/kafka/util/EmbeddedKafka.java:
```diff
@@ -131,28 +138,16 @@ public void createTopics(int partitions, int replication, Properties topicProper
+        properties.setProperty(BOOTSTRAP_SERVERS_CONFIG, getConnectString());
+        properties.setProperty(VALUE_SERIALIZER_CLASS_CONFIG, JsonEncoder.class.getName());
+        properties.setProperty(KEY_SERIALIZER_CLASS_CONFIG, LongSerializer.class.getName());
+        properties.setProperty(PARTITIONER_CLASS_CONFIG, DefaultPartitioner.class.getName());
```
Default partitioner is default. Need to specify?

dcfocus replied:
Fixed. Thanks for catching this one.

dcfocus commented:
> Hi @cricket007 upon tests passed, am mostly good with this PR
> could you please do one more pass before I merge?

Thanks to both. Very helpful comments.
One more change added: the notion `segment` is deprecated. I removed the related code in this commit.

zhenxiao commented on pom.xml:
```diff
@@ -62,6 +62,8 @@
         <dep.nexus-staging-plugin.version>1.6.8</dep.nexus-staging-plugin.version>
         <dep.asm.version>6.2.1</dep.asm.version>
         <dep.gcs.version>1.9.17</dep.gcs.version>
+        <dep.kafka_2.12.version>2.3.1</dep.kafka_2.12.version>
```
kafka.version

zhenxiao commented on pom.xml:
```diff
@@ -62,6 +62,8 @@
         <dep.asm.version>6.2.1</dep.asm.version>
         <dep.gcs.version>1.9.17</dep.gcs.version>
+        <dep.kafka_2.12.version>2.3.1</dep.kafka_2.12.version>
+        <dep.kafka-clients.version>2.3.1</dep.kafka-clients.version>
```
just use kafka.version, no need to create a separate clients version

zhenxiao commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -14,12 +14,12 @@ This connector allows the use of Apache Kafka topics as tables in Presto.
 .. note::
 
-    Apache Kafka 0.8+ is supported although it is highly recommend to use 0.8.1 or later.
+    Latest versions (2.3.1) are used for both kafka-clients and kafka_2.12.
```
Apache kafka 2.3.1+ is supported

zhenxiao commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -104,14 +105,19 @@ timeouts, increasing this value is a good strategy.
+Maximum number of bytes from one partition per poll
 
-This property is optional; the default is ``64kb``.
+This property is optional; the default is ``1048576``.
```
1024Kb * 1024 = 1MB?

zhenxiao commented on pom.xml:
```diff
@@ -1048,9 +1049,75 @@
                 </exclusions>
             </dependency>
 
+            <dependency>
```
could we add this dependency to presto-kafka/pom.xml?

zhenxiao commented:
@dcfocus could you please fix the test failures?

dcfocus commented:
> @dcfocus could you please fix the test failures?

Hi @zhenxiao The maximum run time for each check is limited to 1 hour 20 min. Sometimes it will finish within that limit. It doesn't allow me to re-run. Can you re-run it on your side?

zhenxiao commented on presto-server/pom.xml:
```diff
@@ -31,7 +31,7 @@
         </dependency>
 
-        <!-- The scope of all of the following dependencies are provided to avoid them being added to lib -->
+        <!-- The scope of all of the following dependencies (except Kafka) are provided to avoid them being added to lib -->
```
why are we making these changes? this is for presto-server, better not to pollute the lib

cricket007 commented on presto-kafka/src/test/java/com/facebook/presto/kafka/util/TestUtils.java:
```diff
@@ -71,7 +71,8 @@ public static void installKafkaPlugin(EmbeddedKafka embeddedKafka, QueryRunner q
     {
-        try (KafkaProducer<Long, Object> producer = embeddedKafka.createProducer();
+        try (
+                KafkaProducer<Long, Object> producer = embeddedKafka.createProducer();
```
I see nothing wrong with try with resources
